storage/src/key_value/mod.rs

```diff
@@ -0,0 +1,178 @@
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{any::Any, borrow::Cow, fmt};
+
+use tokio::sync::{mpsc, oneshot};
+
+use crate::{BlockFilter, Digest, SerialBlock, SerialRecord};
+use anyhow::*;
+
+mod storage;
+pub use storage::*;
+
+mod column;
+pub use column::*;
+
+mod agent;
+use agent::Agent;
+
+pub type Value<'a> = Cow<'a, [u8]>;
+
+pub trait KeyValueStorage {
+    fn get<'a>(&'a mut self, column: KeyValueColumn, key: &[u8]) -> Result<Option<Value<'a>>>;
+
+    fn exists(&mut self, column: KeyValueColumn, key: &[u8]) -> Result<bool>;
+
+    fn get_column_keys<'a>(&'a mut self, column: KeyValueColumn) -> Result<Vec<Value<'a>>>;
+
+    fn get_column<'a>(&'a mut self, column: KeyValueColumn) -> Result<Vec<(Value<'a>, Value<'a>)>>;
+
+    fn store(&mut self, column: KeyValueColumn, key: &[u8], value: &[u8]) -> Result<()>;
+
+    fn delete(&mut self, column: KeyValueColumn, key: &[u8]) -> Result<()>;
+
+    fn begin(&mut self) -> Result<()>;
+
+    fn abort(&mut self) -> Result<()>;
+
+    fn commit(&mut self) -> Result<()>;
+
+    fn truncate(&mut self, column: KeyValueColumn) -> Result<()> {
+        let keys = self
+            .get_column_keys(column)?
+            .into_iter()
+            .map(|x| x.into_owned())
+            .collect::<Vec<_>>();
+        for key in keys {
+            self.delete(column, &key[..])?;
+        }
+        Ok(())
+    }
+}
+
+enum Message {
+    InsertBlock(SerialBlock),
+    DeleteBlock(Digest),
+    GetBlockHash(u32),
+    GetBlockHeader(Digest),
+    GetBlockState(Digest),
+    GetBlockStates(Vec<Digest>),
+    GetBlock(Digest),
+    GetForkPath(Digest, usize),
+    CommitBlock(Digest, Digest),
+    DecommitBlocks(Digest),
+    Canon(),
+    LongestChildPath(Digest),
+    GetBlockLocatorHashes(),
+    FindSyncBlocks(Vec<Digest>, usize),
+    GetTransactionLocation(Digest),
+    GetRecordCommitments(Option<usize>),
+    GetRecord(Digest),
+    StoreRecords(Vec<SerialRecord>),
+    GetCommitments(),
+    GetSerialNumbers(),
+    GetMemos(),
+    GetLedgerDigests(),
+    ResetLedger(Vec<Digest>, Vec<Digest>, Vec<Digest>, Vec<Digest>),
+    GetCanonBlocks(Option<u32>),
+    GetBlockHashes(Option<u32>, BlockFilter),
+}
+
+impl fmt::Display for Message {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Message::InsertBlock(block) => write!(f, "InsertBlock({})", block.header.hash()),
+            Message::DeleteBlock(block) => write!(f, "DeleteBlock({})", block),
+            Message::GetBlockHash(block_number) => write!(f, "GetBlockHash({})", block_number),
+            Message::GetBlockHeader(hash) => write!(f, "GetBlockHeader({})", hash),
+            Message::GetBlockState(hash) => write!(f, "GetBlockState({})", hash),
+            Message::GetBlockStates(hashes) => {
+                write!(f, "GetBlockStates(")?;
+                for hash in hashes {
+                    write!(f, "{}, ", hash)?;
```

Comment:
nit (applies to some other places too): unless we do an extra check for the last value, we will have a trailing `, ` right before the `)` in the output string

Reply:
this was a super-quick debug function, not too worried about a trailing comma.